### PR TITLE
adding lidar wrapper

### PR DESCRIPTION
**Pull request by kimmylin727**

### CODE REVIEW

File: SensorCode/SensorState.py [hunk 1]
```diff
@@ -34,12 +34,27 @@ def set_lidar(self, data):
...
-            if j==None:
-                continue
-            ans.append((i,j))
-        return ans
+        vis_map = {}
+        vis_angles = [False] * 360
+        count = 0
+        while count < 360:
+            list_tup = LIDAR_API.get_LIDAR_tuples()
+            for i, j in list_tup:
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: for ang, dist in list_top (Just better variable naming)

File: SensorCode/SensorState.py [hunk 2]
```diff
@@ -34,12 +34,27 @@ def set_lidar(self, data):
...
+                if i not in vis_map:
+                    for x in [-2, -1, 0, 1, 2]:
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: rename x with better variable name, index_offset?

File: SensorCode/SensorState.py [hunk 3]
```diff
@@ -34,12 +34,27 @@ def set_lidar(self, data):
             self.lidar[i] = j
 
     def get_lidar(self):
-        ans= []
-        for i, j in enumerate(self.lidar):
-            if j==None:
-                continue
-            ans.append((i,j))
-        return ans
+        vis_map = {}
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: Definitely add comments describing what these datastructures represent. Since it is not very obvious what they represent

File: SensorCode/SensorState.py [hunk 4]
```diff
@@ -34,12 +34,27 @@ def set_lidar(self, data):
...
+        vis_angles = [False] * 360
+        count = 0
+        while count < 360:
+            list_tup = LIDAR_API.get_LIDAR_tuples()
+            for i, j in list_tup:
+                if i not in vis_map:
+                    for x in [-2, -1, 0, 1, 2]:
+                        if i-x >= 0 and not vis_angles[i-x]:
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: I think you need to handle both the i-x case and i+x case.

Comment (Z01iC) [reply SensorCode/SensorState.py]: Also it is confusing that you are indexing with -x since your list goes -2, -1, 0, 1, 2. Because this means that i--x will actually index i--2 as i+2. Still works tho

File: SensorCode/SensorState.py [hunk 5]
```diff
@@ -34,12 +34,27 @@ def set_lidar(self, data):
...
+                            vis_angles[i-x] = True
+                            count += 1
+            vis_map[i] = j
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: this NEEDS to be tabbed (executed at every iteration of the for loop)

File: SensorCode/SensorState.py (hunk 2 above)
Comment (Z01iC) [reply SensorCode/SensorState.py]: ang_offset?

File: C1C0_path_planning.iml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (ngidwani123) [inline C1C0_path_planning.iml]: Get rid of this file, should not be here.

Comment (Z01iC) [reply C1C0_path_planning.iml]: Amen

File: Grid_Classes/Tile.py [hunk 7]
```diff
@@ -36,11 +36,10 @@ def decrease_score(self, sensor_type):
         If all elements of obstacle_score are at 0, then the tile is considered no longer an obstacle.
         """
         self.obstacle_score[sensor_type] = max(0, self.obstacle_score[sensor_type]-decr_obs_score)
-        if (self.obstacle_score[sensor_type] == 0):
+        if not any(self.obstacle_score):
```
Comment (Z01iC) [inline Grid_Classes/Tile.py:39]: makes sense, but also I think the previous implementation worked as well. Just interested why you made the change? Also delete the random comments that are after this

File: SensorCode/IMU_API.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
+import serial
```
Comment (Z01iC) [inline SensorCode/IMU_API.py:1]: ECE's have some messssy code -_-

File: SensorCode/R2Protocol/R2Protocol.h [hunk 9]
```diff
@@ -0,0 +1,323 @@
+#ifndef R2_PROTOCOL_H
```
Comment (Z01iC) [inline SensorCode/R2Protocol/R2Protocol.h:1]: Why is this file here twice???

File: SensorCode/SensorState.py [hunk 10]
```diff
@@ -1,29 +1,38 @@
...
 
 class SensorState:
     """
     Class to keep track of the state of the sensor inputs for C1C0, to be stored on Jetson
         Instance Attributes:
             lidar (List[tuple[int, int]]): List of lidar values, each element is an (ang,dist) tuple
             terabee_bot (List[int]): List of bottom terabee distances, each element is distance
+            teerabee_mid (List[int]): List of mid terabee distances, each element is distance
             terabee_top (List[int]): List of top terabee distances, each element is distance
             pos_x (int): X position of hedgehog in GPS sub-map
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: Kill the comments about pos_x and pos_y

File: SensorCode/SensorState.py [hunk 11]
```diff
@@ -1,29 +1,38 @@
...
-        self.terabee_bot: List[int] = [1]*20
-        self.terabee_mid: List[int] = [1]*20
-        self.terabee_top: List[int] = [1]*20
+        self.terabee_bot: List[int] = [1]*8
+        self.terabee_mid: List[int] = [1]*8
+        self.terabee_top: List[int] = [1]*8
+        self.imu_gyro: List[int] = [1]*3
+        self.imu_linear_acc: List[int] = [1]*3
         self.heading: int = 0
+        LIDAR_API.init_serial('/dev/ttyTHS1', 38400)
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: Do you need to do this for terabees?

File: ServerGUI.py [hunk 12]
```diff
@@ -92,6 +92,18 @@ def nextLoc(self):
...
+        t_bot, t_mid, t_top = self.sensor_state.get_terabee()
+        returner1 = self.grid.update_grid_tup_data(self.curr_tile.x, self.curr_tile.y, self.sensor_state.get_lidar(),
+                                                   Tile.lidar, robot_radius, bloat_factor, self.path_set)
+        returner2 = self.grid.update_grid_tup_data(self.curr_tile.x, self.curr_tile.y, t_bot, Tile.bottom_terabee,
+                                                   robot_radius, bloat_factor, self.path_set)
+        returner3 = self.grid.update_grid_tup_data(self.curr_tile.x, self.curr_tile.y, t_mid, Tile.mid_terabee,
+                                                   robot_radius, bloat_factor, self.path_set)
+        returner4 = self.grid.update_grid_tup_data(self.curr_tile.x, self.curr_tile.y, t_top, Tile.top_terabee,
+                                                   robot_radius, bloat_factor, self.path_set)
+        return returner1 or returner2 or returner3 or returner4
```
Comment (Z01iC) [inline ServerGUI.py]: returner1 , returner2 ... bad variable naming

File: SensorCode/SensorState.py [hunk 13]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
+                vis_map[ang] = dist
+        # self.lidar = list(vis_map.items())
+        return list(vis_map.items()) #or just do line above and refactor servergui?
+
+        # ans= []
+        # for i, j in enumerate(self.lidar):
+        #     if j==None:
+        #         continue
+        #     ans.append((i,j))
+        # return ans
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: delete comments

File: SensorCode/SensorState.py [hunk 14]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
             self.lidar[i] = j
 
     def get_lidar(self):
```
Comment (Z01iC) [inline SensorCode/SensorState.py:42]: For complicated function write a quick doc string about what it does

File: SensorCode/SensorState.py [hunk 15]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
-                continue
-            ans.append((i,j))
-        return ans
+        vis_map = {} #a dictionary associating angles with object distance
+        vis_angles = [False] * 360 #List of visited angles with a margin of +-2
+        count = 0
+        while count < 360:
+            # list_tup = LIDAR_API.get_LIDAR_tuples()
+            list_tup = TEST_API.get_array("LDR")
+            print(list_tup)
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: kill printssss

File: SensorCode/SensorState.py [hunk 16]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
+                if ang not in vis_map:
+                    for index_offset in [-2, -1, 0, 1, 2]:
+                        #angles with +-2 of read-in angle are also treated as visited
+                        print("test: " + str(ang-index_offset) + " ")
+                        if 0 <= ang-index_offset < 360 and not vis_angles[ang-index_offset]:
+                            vis_angles[ang-index_offset] = True
+                            count += 1
+                vis_map[ang] = dist
+        # self.lidar = list(vis_map.items())
+        return list(vis_map.items()) #or just do line above and refactor servergui?
```
Comment (Z01iC) [inline SensorCode/SensorState.py:60]: Nice use of python functions. Have you verified that the output is the correct type?

File: SensorCode/SensorState.py [hunk 17]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
+        #     ans.append((i,j))
+        # return ans
+
+
+    def get_terabee(self):
+        """
+
+        :return: 3 lists of tuples representing (angle, distance) for bottom, mid, top terabee sensors respectively
+        """
+        returner1 = [] # bottom terabee list of tuples
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: Better naming

File: SensorCode/SensorState.py [hunk 18]
```diff
@@ -48,7 +117,16 @@ def update(self) -> None:
         """
         Update function to read the serial lines and update the sensor state
         """
-        pass
+        self.update_terabee()
```
Comment (Z01iC) [inline SensorCode/SensorState.py:105]: you return terabee data, but never set self.terabee_top, self.terabee_mid ... equal to the returned data

File: SensorCode/SensorState.py [hunk 19]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
+            returner3.append((self.terabee_top_ang[counter], distance))
+            counter = counter + 1
+
+        return returner1, returner2, returner3
+
+
+    def update_terabee(self):
+        # set instance attributes terabee_bot, terabee_mid, and terabee_top to data returned by TERABEE sensor API
+        # self.terabee_bot, self.terabee_mid, self.terabee_top = TERABEE_API.get_terabee_array()
+        self.terabee_bot = TEST_API.get_array("TB1")
```
Comment (Z01iC) [inline SensorCode/SensorState.py:89]: What are the types of these? Just for our reference

File: SensorCode/SensorState.py [hunk 20]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
+        """
+
+        :return: 3 lists of tuples representing (angle, distance) for bottom, mid, top terabee sensors respectively
+        """
+        returner1 = [] # bottom terabee list of tuples
+        returner2 = [] # mid terabee list of tuples
+        returner3 = [] # top terabe list of tuples
+
+        counter = 0
+        for distance in self.terabee_bot:
```
Comment (Z01iC) [inline SensorCode/SensorState.py]: If you want to go the extra mile and do it in two lines instead of 4

for i, distance in zip(range(len(self.terabee_bot)), self.terabee_bot):
    returner1.append((self.terabee_bot_ang[i], distance))

File: SensorCode/SensorState.py [hunk 21]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
-            if j==None:
-                continue
-            ans.append((i,j))
-        return ans
+        vis_map = {} #a dictionary associating angles with object distance
+        vis_angles = [False] * 360 #List of visited angles with a margin of +-2
+        count = 0
+        while count < 360:
+            # list_tup = LIDAR_API.get_LIDAR_tuples()
+            list_tup = TEST_API.get_array("LDR")
```
Comment (Z01iC) [inline SensorCode/SensorState.py:48]: Why do you use TEST_API instead of the actual LIDAR_API, IMU_API etc

File: SensorCode/SensorState.py [hunk 22]
```diff
@@ -34,12 +43,72 @@ def set_lidar(self, data):
...
+    def update_terabee(self):
+        # set instance attributes terabee_bot, terabee_mid, and terabee_top to data returned by TERABEE sensor API
+        # self.terabee_bot, self.terabee_mid, self.terabee_top = TERABEE_API.get_terabee_array()
+        self.terabee_bot = TEST_API.get_array("TB1")
+        self.terabee_mid = TEST_API.get_array("TB2")
+        self.terabee_top = TEST_API.get_array("TB3")
+
+    def get_imu(self):
+        #set instance attributes imu_gyro and imu_linear_acc to data returned by IMU sensor API
+        self.imu_gyro = TEST_API.get_array("IMUG")
```
Comment (Z01iC) [inline SensorCode/SensorState.py:95]: Please clarify with the ECE's that this is the Global Heading, and not the angular acceleration! Message whoever needs to be messaged, maybe in the sensor integration channel

File: SensorCode/SensorState.py (hunk 11 above)
Comment (pdan101) [inline SensorCode/SensorState.py]: Maybe add a comment here to indicate that the port names will change depend on the machine, so we can make sure to change it at the right times

**Review by pdan101 — APPROVED** [review]
Other than my one suggestion for a comment, looks good!

**Review by Z01iC — CHANGES_REQUESTED** [review]
Great work, please make changes accordingly. Super excited to deploy this code and see if the sensors workkkkk! 🚀

File: Grid_Classes/Tile.py (hunk 7 above)
Comment (kimmylin727) [reply Grid_Classes/Tile.py:39]: The original if-statement was I think dummy code to temporarily check if the currently observed sensor is at 0 (which made sense when the only sensor we had to test was lidar), but for setting is_obstacle to false with more sensor data available, we want to make sure that all the sensors are at 0 before toggling the boolean to false.

File: SensorCode/SensorState.py (hunk 11 above)
Comment (kimmylin727) [reply SensorCode/SensorState.py]: yerp

File: SensorCode/R2Protocol.h [hunk 23]
```diff
@@ -0,0 +1,323 @@
...
+  0x26d3, 0x36f2, 0x0691, 0x16b0, 0x6657, 0x7676, 0x4615, 0x5634,
+  0xd94c, 0xc96d, 0xf90e, 0xe92f, 0x99c8, 0x89e9, 0xb98a, 0xa9ab,
+  0x5844, 0x4865, 0x7806, 0x6827, 0x18c0, 0x08e1, 0x3882, 0x28a3,
+  0xcb7d, 0xdb5c, 0xeb3f, 0xfb1e, 0x8bf9, 0x9bd8, 0xabbb, 0xbb9a,
+  0x4a75, 0x5a54, 0x6a37, 0x7a16, 0x0af1, 0x1ad0, 0x2ab3, 0x3a92,
+  0xfd2e, 0xed0f, 0xdd6c, 0xcd4d, 0xbdaa, 0xad8b, 0x9de8, 0x8dc9,
+  0x7c26, 0x6c07, 0x5c64, 0x4c45, 0x3ca2, 0x2c83, 0x1ce0, 0x0cc1,
+  0xef1f, 0xff3e, 0xcf5d, 0xdf7c, 0xaf9b, 0xbfba, 0x8fd9, 0x9ff8,
+  0x6e17, 0x7e36, 0x4e55, 0x5e74, 0x2e93, 0x3eb2, 0x0ed1, 0x1ef0
+};
```
Comment (ngidwani123) [inline SensorCode/R2Protocol.h:42]: What are these values?

File: SensorCode/SensorState.py (hunk 16 above)
Comment (kimmylin727) [reply SensorCode/SensorState.py:60]: yup!!

File: SensorCode/SensorState.py (hunk 18 above)
Comment (kimmylin727) [reply SensorCode/SensorState.py:105]: the function update_terabee() doesn't return anything but instead just sets self.terabee_top, self.terabee_mid to their respective values

File: SensorCode/SensorState.py (hunk 19 above)
Comment (kimmylin727) [reply SensorCode/SensorState.py:89]: They are lists of distances. I'll add a comment clarifying this.

**Review by ngidwani123 — APPROVED** [review]

File: SensorCode/R2Protocol.h (hunk 23 above)
Comment (angelac345) [reply SensorCode/R2Protocol.h:42]: ^^^

**Review by angelac345 — APPROVED** [review]
